### PR TITLE
Webpack optimization

### PR DESCRIPTION
**Pull request by gretzky**
Webpack has come a long way. I think it's in our best interest to slim the webpack config(s) down and see if we can get our build size as small as possible, as well as utilize new webpack features and remove/replace certain plugins.

I'm going to work on the below, and once completed I'll get some opinions from anyone that wants to check it out.

**Removals / migrations**

- [x] Remove the need for the DLL plugin. We can do this now in standalone webpack through `optimization.splitChunks`
- [x] Remove injecting import via webpack import syntax in `app.js` (I'm looking at the favicon, .htaccess, etc.) and migrate those files instead to a `public/` dir along with `index.html` (a la create-react-app) to follow best practice
- [x] ~~Remove extract-text-webpack-plugin in favor of mini-css-extract-plugin [see here](https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/749)~~ Not necessary
- [x] Remove offline-plugin in favor of [sw-precache-webpack-plugin](https://github.com/goldhand/sw-precache-webpack-plugin) for slimmer builds
- [x] Add [assets-webpack-plugin](https://github.com/ztoben/assets-webpack-plugin) to generate hashed bundles for cache busting purposes

**Aesthetics**

- [x] ~~Add [errorOverlayMiddleware](https://github.com/facebook/create-react-app/tree/next/packages/react-error-overlay), again a la cra, for fast display of error messages~~ I think it might be a better idea to migrate away from webpack hot/dev middlewares and utilize proxies
- [x] Add [FriendlyErrorsPlugin](https://github.com/geowarin/friendly-errors-webpack-plugin), for...friendliness
- [x] Remove webpack-bundle-analyzer in favor of [webpack-visualizer](https://github.com/chrisbateman/webpack-visualizer) for a few reasons: with visualizer we can remove the need for a separate `analyze` script, and we can remove the need for the user having to drag n drop the outputted json. Visualizer just spits it into a `stats.html` file, and it's nicer looking 😄 
- [x] Add [webpackbar](https://github.com/nuxt/webpackbar) for a friendlier progress bar and build profiler

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -237,6 +217,7 @@
     "stylelint-config-styled-components": "0.1.1",
     "stylelint-processor-styled-components": "1.3.1",
     "svg-url-loader": "2.3.2",
+    "uglifyjs-webpack-plugin": "^1.2.7",
```
Comment (julienben) [inline package.json]: Don't forget to pin this ;)

**Comment by julienben** [comment]
(FYI: You can now use `dev` as your base.)

**Comment by pouyas-github** [comment]
Do you think this will be merged before next release?

**Comment by julienben** [comment]
That's the plan yes.

Speaking of which, @gretzky, will you have some free time to finish it or would you like someone else to give it a shot?

**Comment by gretzky** [comment]
Howdy @julienben sorry for my absence. I can finish this up this week

**Comment by satishjhanwer** [comment]
@gretzky I am not sure if its an issue or not though I wanted to report just in case. I took checkout of `webpack-optimization` branch, I was testing the HMR as I also wanted to remove `dll plugin` from my current project, it seems HMR is not working correctly. Please find below the steps.
* start the project
* click on `feature` button on `home` page
* now add any new `html` element( let say a `h4` tag with some text) inside the `feature` page and `save`
* check the browser that newly added element is not visible

**Comment by nonotest** [comment]
just a few pointers if anyone is using this PR;
the service worker needs to be passed a config object (with optional onUpdate/onSuccess methods)
there is a type in the service worker filename, needs to be "service-worker" in serviceWorker.js and the webpack config file

**Comment by devmarwen** [comment]
Tried this PR on a project using `css modules`, `sass`

**Previous behaviour**
css updates takes ***200..300ms***

**After applying this PR**
css updates takes ***more than 2s***

Maybe reconsider the removal `DLL Plugin`

**Comment by julienben** [comment]
@gretzky If you're too busy to finish this, let me know and I can give it a shot.

**Comment by gretzky** [comment]
My fault for letting this get out of date-- I'm actually gonna close it bc I believe I have some better solutions for webpack. I'm traveling tomorrow so I'll have some time to blast it out while in airport / flying.

**Comment by julienben** [comment]
Perfect! Happy to help! I think this has the most useful changes for the next version.

**Pull request closed without merging**